server/src/internalClusterTest/java/org/elasticsearch/synonyms/SynonymsManagementAPIServiceIT.java

```diff
@@ -0,0 +1,219 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.synonyms;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.index.mapper.extras.MapperExtrasPlugin;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.reindex.ReindexPlugin;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.junit.Before;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static org.elasticsearch.action.synonyms.SynonymsTestUtils.randomSynonymsSet;
+
+public class SynonymsManagementAPIServiceIT extends ESIntegTestCase {
+
+    private SynonymsManagementAPIService synonymsManagementAPIService;
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return List.of(ReindexPlugin.class, MapperExtrasPlugin.class);
+    }
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        synonymsManagementAPIService = new SynonymsManagementAPIService(client());
+    }
+
+    public void testCreateManySynonyms() throws Exception {
+        CountDownLatch putLatch = new CountDownLatch(1);
+        String synonymSetId = randomIdentifier();
+        int rulesNumber = randomIntBetween(
+            SynonymsManagementAPIService.MAX_SYNONYMS_SETS / 2,
+            SynonymsManagementAPIService.MAX_SYNONYMS_SETS
+        );
+        synonymsManagementAPIService.putSynonymsSet(synonymSetId, randomSynonymsSet(rulesNumber, rulesNumber), new ActionListener<>() {
+            @Override
+            public void onResponse(SynonymsManagementAPIService.SynonymsReloadResult synonymsReloadResult) {
+                assertEquals(
+                    SynonymsManagementAPIService.UpdateSynonymsResultStatus.CREATED,
+                    synonymsReloadResult.synonymsOperationResult()
+                );
+                putLatch.countDown();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                fail(e);
+            }
+        });
+
+        putLatch.await(5, TimeUnit.SECONDS);
+
+        CountDownLatch getLatch = new CountDownLatch(1);
+        // Also retrieve them
+        assertBusy(() -> {
+            synonymsManagementAPIService.getSynonymSetRules(
+                synonymSetId,
+                0,
+                SynonymsManagementAPIService.MAX_SYNONYMS_SETS,
+                new ActionListener<>() {
+                    @Override
+                    public void onResponse(PagedResult<SynonymRule> synonymRulePagedResult) {
+                        // TODO This fails in CI but passes locally. Why?
+                        assertEquals(rulesNumber, synonymRulePagedResult.totalResults());
```

Comment:
I don't understand why this fails in CI. Results should be updated correctly as we're doing a refresh when we update in the API side 🤔

Reply:
Timing?